nemo/collections/asr/models/asr_model.py

```diff
@@ -171,6 +173,51 @@ def on_after_backward(self):
                 logging.warning(f'detected inf or nan values in gradients! Setting gradients to zero.')
                 self.zero_grad()
 
+    @classmethod
+    def disable_cuda_graphs_in_decoder(cls, module: nn.Module):
+        """
+        Disable CUDA graphs for decoding to preserve memory and avoid problems with memory usage.
+        Should be used in training pipeline
+        """
+        # in RNN-T model, `model.decoding` is not an instance of nn.Module,
+        # we need to check `model.decoding.decoding` explicitly
+        for submodule in module.modules():
+            if (
+                hasattr(submodule, "decoding")
+                and hasattr(submodule.decoding, "decoding")
+                and isinstance(submodule.decoding.decoding, WithOptionalCudaGraphs)
+            ):
+                submodule.decoding.decoding.disable_cuda_graphs()
+
+    @classmethod
+    def enable_cuda_graphs_in_decoder(cls, module: nn.Module):
+        """Enable CUDA graphs for decoding (validation/testing)"""
+        # in RNN-T model, model.decoding is not an instance of nn.Module,
+        # we need to check `model.decoding.decoding` explicitly
+        for submodule in module.modules():
+            if (
+                hasattr(submodule, "decoding")
+                and hasattr(submodule.decoding, "decoding")
+                and isinstance(submodule.decoding.decoding, WithOptionalCudaGraphs)
+            ):
+                submodule.decoding.decoding.maybe_enable_cuda_graphs()
+
+    def on_train_epoch_start(self) -> None:
+        """Decoder with CUDA graphs does not release memory, thus we disable it for training epoch"""
+        self.disable_cuda_graphs_in_decoder(self)
+
+    def on_train_epoch_end(self) -> None:
```

Comment:
What about cleanup of on val test predict epoch end ?

Reply:
- `Training`: We want to disable graphs only in training. So for training, we enable it when training starts and disable it when training ends.
- `Val`: validation can be done separately or inside the training loop. We need to disable CUDA graphs after validation to avoid them in training https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#hooks
- `Test`: We do not need to disable CUDA graphs; we need to enable them only since `test` cannot be used in training.
- `Predict`: Same as `test`

I added comments about this logic to the hooks.